src/CodeGen_Hexagon.cpp

```diff
@@ -2734,16 +2739,23 @@ void CodeGen_Hexagon::visit(const Load *op) {
             CodeGen_Posix::visit(op);
             return;
           }
-          int offset = b->value;
-          if (mod_rem.remainder != ((bytes_in_vector() + offset) % bytes_in_vector())) {
+          int offset = (b->value % lanes);
+          // If the offset is negative, it is because we are trying to load something like
+          // f(x-1). In this case, the remainder should be (vector_width + offset) % vector_width
+          // Or in general for any type, it would be (lanes + offset) % lanes.
+          if (mod_rem.remainder != (lanes + offset) % lanes) {
             CodeGen_Posix::visit(op);
             return;
           }
-          int bytes_off = std::abs(offset) * (op->type.bits() / 8);
-          Expr base_low = offset > 0 ? add->a : simplify(add->a - width);
-          Expr base_high = offset > 0 ? simplify(add->a + width) : add->a;
-          Expr ramp_low = Ramp::make(base_low, 1, width);
-          Expr ramp_high = Ramp::make(base_high, 1, width);
+          // offset tells us that we are off by those many elements from the vector
+          // width. For e.g. a value of -1 when we are loading a half-word vector
+          // means we are reading one half-word left of an aligned address.
+          // so we are 1 * (64 / 32) = 2 bytes off (assuming 64 byte mode).
+          int bytes_off = std::abs(offset) * (native_vector_bytes / lanes);
```

Comment:
I think you'd either want to use both div_imp and mod_imp, or / and %. Both are defined such that (a/b)*b + a%b = a, but they don't mix together cleanly.

If you were to use both div_imp and mod_imp, I suspect you'd only ever have to deal with ramps with positive offsets, which would make the logic simpler.

Aside from that, I agree with Pranav that we should leave indentation issues for a separate pass in the hexagon branch.